drivers/ethernet/eth_native_posix.c

```diff
@@ -0,0 +1,268 @@
+#define SYS_LOG_DOMAIN "eth-posix"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_ETH_NATIVE_POSIX_LEVEL
+
+#include <logging/sys_log.h>
+#include <stdio.h>
+
+#include <kernel.h>
+
+#include <stdbool.h>
+#include <errno.h>
+#include <stddef.h>
+
+#include <net/net_pkt.h>
+#include <net/net_core.h>
+#include <net/net_if.h>
+#include <net/ethernet.h>
+
+#include "eth_native_posix_priv.h"
+
+#if defined(CONFIG_NET_L2_ETHERNET)
+#define _ETH_L2_LAYER ETHERNET_L2
+#define _ETH_L2_CTX_TYPE NET_L2_GET_CTX_TYPE(ETHERNET_L2)
+#define _ETH_MTU 1500
+#endif
+
+#define NET_BUF_TIMEOUT MSEC(10)
+
+struct eth_context {
+	u8_t recv[_ETH_MTU + sizeof(struct net_eth_hdr)];
+	u8_t send[_ETH_MTU + sizeof(struct net_eth_hdr)];
+	u8_t mac_addr[6];
+	struct net_linkaddr ll_addr;
+	struct net_if *iface;
+	const char *if_name;
+	int dev_fd;
+	bool init_done;
+	bool status;
+};
+
+NET_STACK_DEFINE(RX_ZETH, eth_rx_stack, CONFIG_ETH_NATIVE_POSIX_RX_STACK_SIZE,
+		 CONFIG_ETH_NATIVE_POSIX_RX_STACK_SIZE);
+static struct k_thread rx_thread_data;
+
+/* TODO: support multiple interfaces */
+static struct eth_context eth_context_data;
+
+static struct eth_context *get_context(struct net_if *iface)
+{
+	return net_if_get_device(iface)->driver_data;
+}
+
+static int eth_send(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct eth_context *ctx = get_context(iface);
+	struct net_buf *frag;
+	int count = 0;
+
+	/* First fragment contains link layer (Ethernet)
+	 * headers so send it first.
+	 */
+	count = net_pkt_ll_reserve(pkt) + pkt->frags->len;
+	memcpy(ctx->send, net_pkt_ll(pkt), count);
+
+	/* Then the remaining data */
+	frag = pkt->frags->frags;
+	while (frag) {
+		memcpy(ctx->send + count, frag->data, frag->len);
+		count += frag->len;
+		frag = frag->frags;
+	}
+
+	net_pkt_unref(pkt);
+
+	eth_write_data(ctx->dev_fd, ctx->send, count);
+
+	return 0;
+}
+
+#if !defined(CONFIG_ETH_NATIVE_POSIX_RANDOM_MAC)
+static int mac_addr_from_str(struct eth_context *ctx, const char *src)
+{
+	unsigned int len, i;
+	char *endptr;
+
+	len = strlen(src);
+	for (i = 0; i < len; i++) {
+		if (!(src[i] >= '0' && src[i] <= '9') &&
+		    !(src[i] >= 'A' && src[i] <= 'F') &&
+		    !(src[i] >= 'a' && src[i] <= 'f') &&
+		    src[i] != ':') {
+			return -EINVAL;
+		}
+	}
+
+	memset(ctx->mac_addr, 0, sizeof(ctx->mac_addr));
+
+	for (i = 0; i < sizeof(ctx->mac_addr); i++) {
+		ctx->mac_addr[i] = strtol(src, &endptr, 16);
+		src = ++endptr;
+	}
+
+	return 0;
+}
+#endif
+
+static int eth_init(struct device *dev)
+{
+	struct eth_context *ctx = dev->driver_data;
+
+	ARG_UNUSED(ctx);
+
+	SYS_LOG_DBG("[%p] dev %p", ctx, dev);
+
+	return 0;
+}
+
+static struct net_linkaddr *eth_get_mac(struct eth_context *ctx)
+{
+	ctx->ll_addr.addr = ctx->mac_addr;
+	ctx->ll_addr.len = sizeof(ctx->mac_addr);
+
+	return &ctx->ll_addr;
+}
+
+static int read_data(struct eth_context *ctx, int fd)
+{
+	struct net_pkt *pkt;
+	struct net_buf *frag;
+	int ret;
+
+	ret = eth_read_data(fd, ctx->recv, sizeof(ctx->recv));
+	if (ret > 0) {
```

Comment:
instead, could you test for error (or 0) and return there?

It would indent less the below code

Reply:
Sure, why not